adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -274,47 +314,83 @@ def _final_callback(self):
         initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
-        for sqr, var, pg in zip(self._state["sqr_avg"], self._state["var_avg"],
-                                self._optimizer.param_groups):
-            sqr, var = max(float(sqr), 0.0), max(float(var), 1e-6)
-            pg["lr"] = (var + sqr) / (var / scale + sqr) * pg["lr"]
+        lr_factors = self._calculate_lr_factors(scale)
+        for lr_factor, pg in zip(lr_factors, self._optimizer.param_groups):
+            pg["lr"] = lr_factor * pg["lr"]
         self._orig_optimizer_step(*args, **kwargs)
         for lr, pg in zip(initial_lr, self._optimizer.param_groups):
             pg["lr"] = lr
         self._state["progress"] += self.gain(scale)
         self._reset_accumulation()
 
-    def zero_grad(self, *args, **kwargs):
-        if self._should_zero_grad:
-            self._orig_optimizer_zero_grad(*args, **kwargs)
-        else:
-            warnings.warn("skipping zero_grad for accumulated gradient")
 
-    def patch_optimizer(self):
-        """
-        Monkey-patch the optimizer's step function with :meth:`AdaScale.step`.
-        """
-        # TODO: detect if the optimizer has already been patched.
+class AdaScale(ScalingRuleBase):
+    """
+    Implements the AdaScale_ algorithm for scaling the learning rate for
+    distributed and large batch size training. Can be used in combination with
+    ``torch.nn.parallel.DistributedDataParallel`` and ``torch.optim.SGD``.
 
-        @functools.wraps(self._optimizer.step)
-        def step_wrapper(optim, *args, **kwargs):
-            return self.step(*args, **kwargs)
+    .. code-block:: python
 
-        @functools.wraps(self._optimizer.zero_grad)
-        def zero_wrapper(optim, *args, **kwargs):
-            return self.zero_grad(*args, **kwargs)
-        self._optimizer.step = MethodType(step_wrapper, self._optimizer)
-        self._optimizer.zero_grad = MethodType(zero_wrapper, self._optimizer)
+        optim = torch.optim.SGD(model, lr=0.001)
+        model = DistributedDataParallel(model, optim, scaling_rule="AdaScale")
+        adascale = model.scaling_rule
+
+        for epoch in ...:
+            for batch in ...:
+                optim.zero_grad()
+                loss = ...
+                loss.backward()
+                adascale.step()
+
+    .. _AdaScale: https://proceedings.icml.cc/static/paper_files/icml/2020/4682-Supplemental.pdf
+    """  # noqa: E501
+
+    def _calculate_lr_factors(self, scale):
+        """Calculate factors to be applied to lr for each parameter group."""
+        lr_factors = []
+        for sqr, var, pg in zip(self._state["sqr_avg"], self._state["var_avg"],
+                                self._optimizer.param_groups):
+            sqr, var = max(float(sqr), 0.0), max(float(var), 1e-6)
+            lr_factors.append((var + sqr) / (var / scale + sqr))
+        return lr_factors
+
+    def calculate_lr_factor_estimation(self, scale):
```

Comment:
I think `_calculate_lr_factors` and `calculate_lr_factor_estimation` are named a bit too similar. Without reading their body's here, its difficult to figure out what they do and how they are different.

Reply:
any suggestions?

Reply:
I don't think this method really makes sense to be in the lr scaler class. Since it's just for progress tracking, which is independent of lr scaling.

Reply:
This method is used to calculate the `Learning_Rate_Factor` scaler logged on Tensorboard, different from the lr_factor vector which is calculated per parameter group. It looks to me more like a single value estimation of the LR factor applied for this step. Not sure how it can be outside of scaling rules since it is calculated differently for each of them.

Reply:
Oh, if it's just for the visualization, you can just take the average of the individual lr factors. No need to demand users implement custom logic for this.